playbook/actions/gitops.go

```diff
@@ -0,0 +1,232 @@
+	}
+
+	if t.shouldCreatePR {
+		prNumber, err := t.createPR(ctx, connector, workTree)
+		if err != nil {
+			return nil, err
+		}
+		response.CreatedPR = prNumber
+	}
+
+	return &response, nil
+}
+
+// generateSpec generates the spec for the git client from the action
+func (t *GitOps) generateSpec(ctx context.Context, action v1.GitOpsAction) error {
+	if action.Repo.Base == "" {
+		action.Repo.Base = "main"
+	}
+
+	if action.Repo.Branch == "" {
+		action.Repo.Branch = action.Repo.Base
+	}
+
+	templater := gomplate.StructTemplater{
+		RequiredTag: "template",
+		DelimSets: []gomplate.Delims{
+			{Left: "{{", Right: "}}"},
+			{Left: "$(", Right: ")"},
+		},
+		ValueFunctions: true,
+		Values:         t.env.AsMap(),
+	}
+	if err := templater.Walk(&action); err != nil {
+		return fmt.Errorf("failed to walk template: %w", err)
+	}
+
+	t.spec = &connectors.GitopsAPISpec{
+		Repository: action.Repo.URL,
+		Base:       action.Repo.Base,
+		Branch:     action.Repo.Branch,
+		CommitMsg:  action.Commit.Message,
+		User:       action.Commit.AuthorName,
+		Email:      action.Commit.AuthorEmail,
+	}
+
+	if action.Repo.Connection != "" {
+		conn, err := ctx.HydrateConnectionByURL(action.Repo.Connection)
+		if err != nil {
+			return err
+		}
+
+		switch conn.Type {
+		case models.ConnectionTypeGithub:
+			t.spec.GITHUB_TOKEN = conn.Password
+			t.shouldCreatePR = true
+
+		case models.ConnectionTypeAzureDevops:
+			// TODO: Azure devops connection doesn't have git credentials ...?
+
+		case models.ConnectionTypeGit:
+			// TODO: need to finalize this once git connection is implemented
+
+		default:
+			return fmt.Errorf("unsupported connection type: %s", conn.Type)
+		}
+	}
+
+	if action.PullRequest == nil {
+		t.shouldCreatePR = false
+	}
+
+	if t.shouldCreatePR && action.Repo.Base == action.Repo.Branch {
+		logger.Warnf("no base branch was provided on gitops action. So no PR will be created")
+		t.shouldCreatePR = false
+	}
+
+	if t.shouldCreatePR {
+		t.spec.PullRequest = &connectors.PullRequestTemplate{
+			Base:   action.Repo.Base,
+			Branch: action.Repo.Branch,
+			Title:  action.PullRequest.Title,
+			Tags:   action.PullRequest.Tags,
+		}
+	}
+
+	return nil
+}
+
+func (t *GitOps) cloneRepo(ctx context.Context, action v1.GitOpsAction) (connectors.Connector, *gitv5.Worktree, error) {
+	connector, workTree, err := git.Clone(ctx, t.spec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return connector, workTree, nil
+}
+
+func (t *GitOps) applyPatches(ctx context.Context, action v1.GitOpsAction) error {
+	for _, patch := range action.Patches {
+		fullpath := filepath.Join(t.workTree.Filesystem.Root(), patch.Path)
+		paths, err := files.UnfoldGlobs(fullpath)
+		if err != nil {
+			return err
+		}
+
+		for _, path := range paths {
+			relativePath, err := filepath.Rel(t.workTree.Filesystem.Root(), path)
+			if err != nil {
+				return fmt.Errorf("failed to get relative path: %w", err)
+			}
+			logger.Debugf("Patching %s", relativePath)
```

Comment:
These logs should be GitOps object/context so that they can be saved to the db and shown in the ui